### PR TITLE
[SPARK-43123][SQL] Internal field metadata should not be leaked to catalogs

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

In Spark, we have defined some internal field metadata to help query resolution and compilation. For example, there are quite some field metadata that are related to metadata columns.

However, when we create tables, these internal field metadata can be leaked. This PR updates CTAS/RTAS commands to remove these internal field metadata before creating tables. CREATE/REPLACE TABLE command is fine as users can't generate these internal field metadata via the type string.

### Why are the changes needed?

to avoid potential issues, like mistakenly treating a data column as metadata column

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gengliangwang

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala [hunk 1]
```diff
@@ -225,4 +226,20 @@ package object util extends Logging {
       }
     }
   }
+
+  val AUTO_GENERATED_ALIAS = "__autoGeneratedAlias"
+
+  def removeInternalMetadata(schema: StructType): StructType = {
+    StructType(schema.map { field =>
+      val newMetadata = new MetadataBuilder().withMetadata(field.metadata)
+        .remove(METADATA_COL_ATTR_KEY)
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:235]: Nit: shalll we define an array for all the internal metadata
E.g.
```
val internalMetaData = Seq(
  METADATA_COL_ATTR_KEY,
  QUALIFIED_ACCESS_ONLY,
  ...
```

**Review by gengliangwang — APPROVED** [review]
Nice catch

**Comment by cloud-fan** [comment]
thanks for review, merging to master/3.4!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala [hunk 2]
```diff
@@ -225,4 +226,25 @@ package object util extends Logging {
       }
     }
   }
+
+  val AUTO_GENERATED_ALIAS = "__autoGeneratedAlias"
+
+  val INTERNAL_METADATA_KEYS = Seq(
+    AUTO_GENERATED_ALIAS,
```
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:233]: Are these metadata keys only in the top level columns?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:233]: I believe so, after checking the code that generates them. For example, https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala#L171-L175

**Pull request closed without merging**